Add Guide for Apache Pinot

sharon2719 opened the pull request:
closes: https://github.com/apache/airflow/issues/8194
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

sharon2719 committed:
pinot operators guide

theodesp commented on docs/apache-airflow-providers-apache-beam/operators.rst:
```diff
@@ -84,7 +85,7 @@ Run Java Pipelines in Apache Beam
 :class:`~airflow.providers.apache.beam.operators.beam.BeamRunJavaPipelineOperator`
 as it contains the pipeline to be executed by Apache Beam. The JAR can be available on GCS that Airflow
-has the ability to download or available on the local filesystem (provide the absolute path to it).
+has the ability to download or available on  the local filesystem (provide the absolute path to it).
```
```suggestion
has the ability to download or available on the local filesystem (provide the absolute path to it).
```

theodesp commented on docs/apache-airflow-providers-apache-pinot/pinot-operators.rst:
```diff
@@ -0,0 +1,140 @@
+
+`Apache Pinot <https://pinot.apache.org/>`__ is a column-oriented, open-source,
+distributed data store written in Java. Pinot is designed to execute OLAP queries
+with low latency. It is suited in contexts where fast analytics,such as aggregations,
```
```suggestion
with low latency. It is suited in contexts where fast analytics, such as aggregations,
```

sharon2719 committed:
Update docs/apache-airflow-providers-apache-pinot/pinot-operators.rst

Co-authored-by: Theofanis Despoudis <328805+theodesp@users.noreply.github.com>

potiuk committed:
Update docs/apache-airflow-providers-apache-beam/operators.rst

Co-authored-by: Theofanis Despoudis <328805+theodesp@users.noreply.github.com>

sharon2719 committed:
pinot operators guide

sharon2719 committed:
new updates

eladkal commented on docs/apache-airflow-providers-apache-beam/operators.rst:
```diff
@@ -76,6 +76,7 @@ Python Pipelines with DataflowRunner
     :start-after: [START howto_operator_start_python_dataflow_runner_pipeline_async_gcs_file]
     :end-before: [END howto_operator_start_python_dataflow_runner_pipeline_async_gcs_file]
 
+
```
Please revert this addition of new line

eladkal replied:
```suggestion
```

eladkal commented on docs/apache-airflow/concepts/operators.rst:
```diff
@@ -16,7 +16,7 @@
 
 Operators
-=========
+==========
```
Please revert this change

eladkal replied:
```suggestion
=========
```

eladkal commented on docs/apache-airflow-providers-apache-pinot/pinot-operators.rst:
```diff
@@ -0,0 +1,139 @@
+if (arguments.size() < 2) {
+throw new IllegalArgumentException("At least 2 arguments are required for ADD transform function");
+}
+for (TransformFunction argument : arguments) {
```
The guide should be about the Airflow operators/hooks.
Specifically: `PinotDbApiHook`  / `PinotAdminHook` 
https://github.com/apache/airflow/blob/main/airflow/providers/apache/pinot/hooks/pinot.py

sharon2719 committed:
pinot guide

sharon2719 committed:
pinot guide

eladkal committed:
Update docs/apache-airflow-providers-apache-beam/operators.rst

eladkal committed:
Update docs/apache-airflow/concepts/operators.rst

eladkal commented on docs/apache-airflow-providers-apache-pinot/operators.rst:
```diff
@@ -0,0 +1,75 @@
+
+Reference
+^^^^^^^^^
+For more information,please see the documentation at `Apache Pinot improvements for PinotAdminHook<https://github.com/apache/incubator-pinot/pull/4110>`
```
Why do we link to a PR and not to official documentation?

josh-fell replied:
Another small nit for a space between "information,please"

eladkal commented on docs/apache-airflow-providers-apache-pinot/operators.rst:
```diff
@@ -0,0 +1,75 @@
+PinotDbApiHook
+--------------
+
+This hook uses standard-SQL endpoint since PQL endpoint is soon to be deprecated. 
```
```suggestion
This hook uses standard-SQL endpoint. PQL endpoint is removed in updated versions of pinotdb library. For more information see `library documentation  <https://docs.pinot.apache.org/users/clients/python>`
```

eladkal commented on airflow/providers/apache/pinot/example_dags/example_pinot_dag.py:
```diff
@@ -0,0 +1,49 @@
+
+from airflow import DAG
+from airflow.providers.apache.pinot.hooks.pinot import PinotAdminHook
+from airflow.providers.apache.pinot.hooks.pinot import PinotDbApiHook
```
```suggestion
from airflow.providers.apache.pinot.hooks.pinot import PinotAdminHook, PinotDbApiHook
```

eladkal reviewed (approved):
Overall LGTM :) 
Left some comments to address

@josh-fell can you take a look at the example DAG?

sharon2719 committed:
Update airflow/providers/apache/pinot/example_dags/example_pinot_dag.py

Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>

sharon2719 committed:
Update docs/apache-airflow-providers-apache-pinot/operators.rst

Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>

josh-fell commented on airflow/providers/apache/pinot/example_dags/example_pinot_dag.py:
```diff
@@ -0,0 +1,48 @@
+with DAG(
+    dag_id='example_pinot_hook',
+    schedule_interval=None,
+    start_date=days_ago(2),
```
We are currently transitioning all of the example DAGs (and later docs) away from using `days_ago(n)` for `start_date` to a static value since it's best practice when writing pipelines. New examples should use a static `start_date` value.  The updates have been using `datetime()`; the actual value doesn't matter.

josh-fell commented on airflow/providers/apache/pinot/example_dags/example_pinot_dag.py:
```diff
@@ -0,0 +1,48 @@
+        task_id="run_example_pinot_script",
+        pinot="ls /;",
+        pinot_options="-x local",
+        dag=dag,
```
The `dag=dag` can be removed since the `DAG` object is being used as a context manager which takes care of passing this arg to all operators.

josh-fell replied:
Would it be possible to show different args for `pinot` and `pinot_options` parameters? Otherwise, seeing as the same args are used for every operator in the DAG, these ideally should be present as `default_args` at the DAG level. However, because the operator documentation is referencing the example DAG, moving the args to `default_args` loses the main context for showcasing the operators in the documentation.

Updating the args to be different between the two would preclude the `default_args` use.  I'm not sure if they need to be the same value when using these operators together in a pipeline though.

josh-fell commented on docs/apache-airflow-providers-apache-pinot/operators.rst:
```diff
@@ -0,0 +1,75 @@
+Apache Pinot Hooks
+==================
+ 
+`Apache Pinot <https://pinot.apache.org/>`__ is a column-oriented, open-source,distributed data store written in Java. Pinot is designed to execute OLAP querieswith low latency. It is suited in contexts where fast analytics, such as aggregations,are needed on immutable data, possibly, with real-time data ingestion.
```
Small nit for spaces:
> ... a column-oriented, open-source,distributed data store written in Java

"... column-oriented, open-source, distributed data store written in Java"

> ... such as aggregations,are needed

"... such as aggregations, are needed ...

josh-fell commented on docs/apache-airflow-providers-apache-pinot/operators.rst:
```diff
@@ -0,0 +1,75 @@
+Reference
+^^^^^^^^^
+
+For more information, please see the documentation at `Pinot documentation on querrying data <https://docs.pinot.apache.org/users/api/querying-pinot-using-standard-sql>`
```
"querrying" should be spelled "querying"

sharon2719 committed:
pinot updates

eladkal commented on docs/apache-airflow-providers-apache-pinot/operators.rst:
```diff
@@ -0,0 +1,74 @@
+Apache Pinot Hooks
+==================
+
+`Apache Pinot <https://pinot.apache.org/>`__ is a column-oriented, open-source,distributed data store written in Java. Pinot is designed to execute OLAP querieswith low latency. It is suited in contexts where fast analytics, such as aggregations,are needed on immutable data, possibly, with real-time data ingestion.
```
```suggestion
`Apache Pinot <https://pinot.apache.org/>`__ is a column-oriented, open-source, distributed data store written in Java. Pinot is designed to execute OLAP queries with low latency. It is suited in contexts where fast analytics, such as aggregations are needed on immutable data, possibly, with real-time data ingestion.
```

eladkal reviewed (changes requested):
There is an error that needs to be fixed:
```
  "/opt/airflow/airflow/providers/apache/pinot/example_dags/example_pinot_dag.py",
  line 28, in <module>
      start_date=datetime(),
  TypeError: Required argument 'year' (pos 1) not found
```

eladkal committed:
Update docs/apache-airflow-providers-apache-pinot/operators.rst

sharon2719 committed:
new apache updates

sharon2719 committed:
Merge branch 'main' into main

potiuk committed:
Workaround libstdcpp TLS error (#19010)

Workaround https://github.com/apache/airflow/issues/17546 issue with
/usr/lib/x86_64-linux-gnu/libstdc++.so.6: cannot allocate memory in
static TLS block.

We do not yet a more "correct" solution to the problem but in order to
avoid raising new issues by users of the prod image, we implement the
workaround now.

The side effect of this is slightly (in the range of 100s of
milliseconds) slower load for any binary started and a little memory
used for Heap allocated by initialization of libstdc++.

This overhead is not happening for binaries that already link
dynamically libstdc++.

raphaelauv committed:
Google provider catch invalid secret name (#18790)

bbovenzi committed:
Rename execution date in forms and tables (#19063)

pgagnon committed:
Document `hdfs_namenode_principal` for HDFS connections (#18987)

* Document `hdfs_namenode_principal` for HDFS connections

bbovenzi committed:
Fix: Add taskgroup tooltip to graph view (#19083)

uranusjr committed:
Relax packaging requirement (#19087)

fernandobugni committed:
Add docker-compose explanation to conn localhost (#19076)

jedcunningham committed:
Allow Param to support a default value of ``None`` (#19034)

sdebruyn committed:
Add dataroots to Airflow users (#19074)

uranusjr committed:
Rename trigger page label to Logical Date (#19061)

josh-fell committed:
Updating TaskGroup unit test to handle new tooltip (#19089)

SamWheating committed:
Adding dag_id_pattern parameter to the /dags endpoint (#18924)

ephraimbuddy committed:
Ensure task state doesn't change when marked as failed/success/skipped (#19095)

Currently, when a dagrun is marked as success in the UI, the expected
sigterm is sent but the task continues running, changing the dagrun to running
before eventually failing. Same as well when marked as failed.

Also, queued tasks continue running even when the dagrun was failed

The same thing happens when a dagrun times out. The tasks that are marked
skipped, starts again, and then fails.

This PR fixes these issues

uranusjr committed:
Change `ds`, `ts`, etc. back to use logical date (#19088)

OyinOlamide committed:
Fix wrong commands in docs/breeze

ashb committed:
Don't install SQLAlchemy/Pendulum adapters for other DBs (#18745)

This stops the MySQL libs being imported "unnecessarily" when Postgres
is in use -- and there have been a few confusing reports of the mysql
client libs causing problems in rare cases, so lets avoid the import if
we can.

ephraimbuddy committed:
Update taskinstance REST API schema to include dag_run_id field (#19105)

This PR adds dag_run_id field to taskinstance schema

josh-fell committed:
Static start_date and default arg cleanup for Microsoft providers example DAGs (#19062)

ephraimbuddy committed:
Prevent scheduler crash when serialized dag is missing (#19113)

Scheduler._send_dag_callbacks_to_processor calls dag_run.get_dag which
raises exception. This PR changes to calling dagbag.get_dag and changing
Scheduler._send_dag_callbacks_to_processor args to accept dag instead of dag_run.

ephraimbuddy committed:
Fix catchup by limiting queued dagrun creation using max_active_runs (#18897)

Currently, when catchup is True, we create a lot of dagruns limited by
max_queued_runs_per_dag setting. This is not efficient as some dagruns takes
longer to run.

This PR brings back the old behavior of not creating dagruns once max_active_runs
is reached thereby solving the catchup issue.

Now, the dagruns appears as though they were created in running state

prabhjyotsingh committed:
[Minor] Fix padding on home page (#19025)

* [Minor] fix padding on home page

* fix lint

* Move filters and pagination out from dags-table-wrap

* - fix last row of a table
 - col span should be 10 instead of 9

* add bottom padding

* Update airflow/www/static/css/dags.css

Co-authored-by: Ryan Hamilton <ryan@ryanahamilton.com>

Co-authored-by: Ryan Hamilton <ryan@ryanahamilton.com>

bbovenzi committed:
Hide tooltip when next run is none (#19112)

* Hide Next Run when it is None

Check for `is not none` instead of `is defined` when showing the next run tooltip

* check next_dagrun is defined and not none

Acehaidrey committed:
Add Pinterest to Airflow users list (#19117)

mik-laj committed:
Warn about unsupported Python 3.10 (#19060)

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

jedcunningham committed:
Update "Release Airflow" doc (#19111)

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Commit:
Upgrade the Dataproc package to 3.0.0 and migrate from v1beta2 to v1 api (#18879)

uranusjr committed:
Remove distutils usages for Python 3.10 (#19064)

uranusjr committed:
Move away from legacy importlib.resources API (#19091)

ysk24ok committed:
Update docstring to let users use `node_selector` (#19057)

mariotaddeucci committed:
Add sensor default timeout config (#19119)

o-nikolas committed:
Update virtualenv guide to use correct constraints file (#19141)

The constraints file that was used in the example code snippets
is a constraints file for released versions (e.g.:
airflow/constraints-main/constraints-3.6.txt). Which was leading to long
dependency resolution times as well as dependency version conflicts when
using the latest source from main. The correct constraints to use in this
case is for example: airflow/constraints-main/constraints-source-providers-3.6.txt

blag committed:
Fixup string concatenations (#19099)

jedcunningham committed:
Chart: Fix `extraEnvFrom` examples (#19144)

robinedwards committed:
Fix queued dag runs changes catchup=False behaviour (#19130)

jhtimmins committed:
Consolidate method names between Airflow Security Manager and FAB default (#18726)

* Replace add_view_menu with create_resource.

* Replace add_permission_view_menu with create_permission.

* Replace find_permission_view_menu with get_permission.

* Replace find_view_menu with get_resource.

* Replace get_all_view_menu with get_all_resources.

* Replace find_permission with get_action.

* Replace del_permission_view_menu with delete_permission.

* Replace del_view_menu with delete_resource.

* Replace del_permission with delete_action.

* Replace add_permission_role with add_permission_to_role.

* Replace find_permissions_view_menu with get_resource_permissions.

* Replace self.del_permission_role wth remove_permission_from_role.

* Replace exist_permission_on_roles with permission_exists_in_one_or_more_roles.

* Replace find_roles_permission_view_menus with filter_roles_by_perm_with_action.

* Replace get_db_role_permissions with get_role_permissions_from_db.

* Replace add_permission with create_action.

* Remove unused exist_permission_on_view function.

* Rename local variables.

* Remove sqla model names from SecurityManager base class.

* Update names in add_permissions_view.

* Flake8.

* Use updated perm names.

* Use updated perm names for local vars.

* Use black.

* Reorder import statements.

* Remove accidental renaming of BaseSecurityManager.roles.

uranusjr committed:
Crerate TI context with data interval compat layer (#19148)

uranusjr committed:
Add test for interval timetable catchup=False (#19145)

steveyz-astro committed:
add detailed information to logging when a dag or a task finishes. (#19097)

* add detailed information to logging when a dag or a task finishes.

* make logging of start_date/end_date ISO format to be consist

* fix pre-commit

* use only %s in new logging statements to gracefully handle when certain variables are None

* fix precommit

* use self._state instead of get_state().  add computation for dag run duration based on start_date and end_date

* make linter happy with format

* fix typo

* put back missing reference to _state

Co-authored-by: Daniel Imberman <daniel.imberman@gmail.com>

edwardwang888 committed:
Navigate directly to DAG when selecting from search typeahead list (#18991)

When searching for a DAG, we are given a typeahead list of matching DAGs. When selecting a DAG from the typeahead list, the previous behavior was to execute a search query on that dag_id. Instead, we now go directly to that DAG.

josh-fell committed:
Replacing non-attribute template_fields for BigQueryToMsSqlOperator (#19052)

* Replacing non-attribute template_fields for BigQueryToMsSqlOperator

* Updating source_project_dataset_table arg in example DAG

Harisonm committed:
Update INTWILD.md to add Retailink (#19115)

Wadenschwinger committed:
Added Wayfair to the list of companies using Apache Airflow (#19162)

GeoloeG-IsT committed:
Adds Sprylab as user "in the wild" (#19159)

mik-laj committed:
Add TriggererJob to jobs check command (#19179)

o-nikolas committed:
Allow specifying extras when using breeze initialize_local_virtualenv (#19178)

brain-buster committed:
Remove incorrect type comment in Swagger2Specification._set_defaults classmethod (#19065)

o-nikolas committed:
Fix breeze docker version parsing (#19182)

simplylizz committed:
Add migration job resources (#19175)

* Add migration job resources

* Add change notes

* Fix test

* Linter

* Update chart/values.schema.json

* Update chart/values.schema.json

Co-authored-by: Kamil Breguła <mik-laj@users.noreply.github.com>

* Linter

Co-authored-by: Kamil Breguła <mik-laj@users.noreply.github.com>

mik-laj committed:
Fix airflow jobs check cmd for TriggererJob (#19185)

jedcunningham committed:
Faster PostgreSQL db migration to Airflow 2.2 (#19166)

Bigger Airflow databases can take a long time to migrate the database,
particularly if they have a lot of task instances. On PostgreSQL, creating a
new table is much faster than updating the existing table.

ReadytoRocc committed:
Clear ti.next_method and ti.next_kwargs on task finish (#19183)

Acehaidrey committed:
Add role export/import to cli tools (#18916)

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

msumit committed:
Add test_connection method for Snowflake Hook (#19041)

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

kibablu committed:
Add a fix on Ubuntu 20.04 for mysql_config not found error (#19184)

uranusjr committed:
Add DagRun.logical_date as a property (#19198)

uranusjr committed:
Fix static checks (#19200)

joeknize-bc committed:
Moving the example tag a little bit up to include the part where you specify the snowflake_conn_id (#19180)

kimyen committed:
Allow Airflow UI to create worker pod via Clear > Run (#18272)

benjamin-chang committed:
Update to correctly resolve 'PostgresOperator' (#19212)

Changed ``airflow.providers.postgres.operator.postgres.PostgresOperator`` 
to ``airflow.providers.postgres.operators.postgres.PostgresOperator``

ephraimbuddy committed:
Fix Unexpected commit error in schedulerjob (#19213)

keze committed:
Fix bug in Dataflow hook when no jobs are returned (#18981)

potiuk committed:
Removes hard-coding of /tmp filesystem cache (#19208)

Filesystem cache has been hard-coded to /tmp. This would not
work for Windows but (more importantly) it does not work in case
the user has no access to /tmp directory and TMPDIR is set to
a different value.

Fixes: #19206

pgvishnuram committed:
Chart: Add labels to jobs created by cleanup pods (#19225)

Problem:
Currently pods created by cleanup jobs doesn't add labels similar to metadata.labels.
This prevents us to use network policies to block / allow to communication to internal services

How does this PR fix the problem above:

This PR adds ability to utilise existing labels and allow to use new labels from values.yaml

KianYang-Lee committed:
Doc: Improve tutorial documentation and code (#19186)

1. Added instructions on adding postgres connection.
2. Modified proper SQL syntax.
3. Remove redundant lines when writing to CSV.
4. Added QUOTE argument for copy_expert

potiuk committed:
Fix hard-coded /tmp directory in CloudSQL Hook (#19229)

The /tmp directory is not "guaranteed" to be THE /tmp one. There
are cases where TMPDIR or other env variables are set to override
it (for example when user has no write access to /tmp dir) and
we should respect that. The gettempdir() will produce the right
temporary directory based on the env variables set in the system.

sharon2719 committed:
Apache Pinot updates

sharon2719 reviewed (commented):
Changed already

sharon2719 reviewed (commented):
Changed already

sharon2719 committed:
Merge branch 'main' into Resove_conflicts

sharon2719 committed:
Merge branch 'apache:main' into Resove_conflicts
